cpp/src/arrow/device.h

```diff
@@ -105,6 +107,63 @@ class ARROW_EXPORT Device : public std::enable_shared_from_this<Device>,
   bool is_cpu_;
 };
 
+/// \brief EXPERIMENTAL: An object that provides event/stream sync primitives
+///
+///
+class ARROW_EXPORT DeviceSync {
+ public:
+  explicit DeviceSync(void* sync_event) : sync_event_{sync_event}, owns_event_{false} {}
+  virtual ~DeviceSync() = default;
+
+  /// @brief Block until sync event is completed.
+  ///
+  /// Should be a no-op for CPU devices.
+  virtual Status wait() = 0;
+
+  /// @brief Make the provided stream wait on the sync event.
+  ///
+  /// Tells the provided stream that it should wait until the
+  /// synchronization event is completed without blocking the CPU.
+  /// @param stream Should be appropriate for the underlying device
+  virtual Status stream_wait(void* stream) = 0;
+
+  virtual void set_stream(void* stream) { stream_ = stream; }
+
+  Result<void*> get_event() {
+    if (!sync_event_) {
+      ARROW_ASSIGN_OR_RAISE(sync_event_, create_event());
+      owns_event_ = true;
+    }
+    return sync_event_;
+  }
+
+  void clear_event() {
```

Comment:
Would this be used (instead of just destroying the DeviceSync)? I think it'd be best to keep the interface minimal until we can document the usage which each function anticipates

Reply:
Yes, this has to be used instead of just destroying the DeviceSync for the same reason as above, it's not safe to call a polymorphic virtual function from a destructor.